README.md
# SMLReader

An ESP8266 based smart meter (SML) to MQTT gateway 

## About

The aim of this project is to read the meter readings of modern energy meters and make them available via MQTT.

The software was primarily developed and tested for the EMH ED300L electricity meter, but should also work with other energy meters that have an optical interface and communicate via the SML protocol.

SMLReader publishes the metrics read from the meter's optical unit to an MQTT broker configured via the provided web interface.

If you like this project, you might consider to [support me](#donate).

### Screenshots
![WiFi and MQTT setup](doc/screenshots/screenshot_setup.png)
![Grafana](doc/screenshots/screenshot_grafana.png)
```bash
MB-Monty ➜  ~  mosquitto_sub -h 10.4.32.103 -v -t smartmeter/mains/#
smartmeter/mains/info Hello from 00C7551E, running SMLReader version 2.1.5.
smartmeter/mains/sensor/1/obis/1-0:1.8.0/255/value 3546245.9
smartmeter/mains/sensor/1/obis/1-0:2.8.0/255/value 13.2
smartmeter/mains/sensor/1/obis/1-0:1.8.1/255/value 0.0
smartmeter/mains/sensor/1/obis/1-0:2.8.1/255/value 13.2
smartmeter/mains/sensor/1/obis/1-0:1.8.2/255/value 3546245.9
smartmeter/mains/sensor/1/obis/1-0:2.8.2/255/value 0.0
smartmeter/mains/sensor/1/obis/1-0:16.7.0/255/value 451.2
```

### Hardware

#### Reading head

The reading head consists of a phototransistor (BPW 40) and a 1 kΩ pull-up resistor connected to one of the GPIO pins of the microcontroller.
Other phototransistors or the use of an internal pull-up resistor will probably work, too.

The housing of my reading head has been 3D-printed using the [STL files](http://www.stefan-weigert.de/php_loader/sml.php) from [Stefan Weigert](http://www.stefan-weigert.de). 

A ring magnet (in my case dimensioned 27x21x3mm) ensures that the reading head keeps stuck on the meter.

The phototransistor has been fixed with hot glue within the housing.

![Reading Head](doc/assets/SMLReader_Img_ReadingHead.jpg "Reading Head") ![Reading Head](doc/assets/SMLReader_Img_ReadingHead_Close.jpg "Reading Head") ![D1](doc/assets/SMLReader_Img_D1.jpg "WeMos D1 mini")

#### Schematic diagram
![Schematic diagram](doc/assets/SMLReader_Schema.png)

## Getting started

To get started, the software must first somehow get onto the device. 
This can be done in several ways.

### Uploading a precompiled binary

A precompiled binary for the Wemos D1 and configured to have a sensor attached to pin D2 can be downloaded from the releases page on GitHub:  

[https://github.com/mruettgers/SMLReader/releases](https://github.com/mruettgers/SMLReader/releases)

#### Flashing

This precompiled binary can be flashed to the device with the help of [esptool.py](https://github.com/espressif/esptool).
Other tools should also work, as long as you have configured them to match the specs of your ESP8266.

```bash
esptool.py --port /dev/ttyUSB0 write_flash -fm dout 0x00000 path/to/smlreader.bin
```

As a docker enthusiast I prefer to utilize a dockerized version of esptool.py by bind-mounting the current working directory to `/src` and assuming that `smlreader.bin` is existing in the current working directory:
```bash
docker run -it --device /dev/ttyUSB0 -v $(pwd):/src --rm mruettgers/esptool --port /dev/ttyUSB0 write_flash -fm dout 0x00000 /src/smlreader.bin
```

The device path and the path to the binary have to be adjusted to fit your needs.


##### Example

```bash
nb-ubuntu ➜  ~/Downloads  ls SMLReader_D1mini_v2.1.3.bin                                                                                                         
SMLReader_D1mini_v2.1.3.bin

nb-ubuntu ➜  ~/Downloads  docker run -it --device=/dev/ttyUSB0 -v $(pwd):/src --rm mruettgers/esptool --port /dev/ttyUSB0 write_flash -fm dout 0x00000 /src/SMLReader_D1mini_v2.1.3.bin
esptool.py v2.8
Serial port /dev/ttyUSB0
Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: d8:f1:5b:07:0d:eb
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Auto-detected Flash size: 4MB
Flash params set to 0x0340
Compressed 378528 bytes to 266738...
Wrote 378528 bytes (266738 compressed) at 0x00000000 in 23.6 seconds (effective 128.1 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
```

---

### Using your IDE for building and flashing

You should be able to use your preferred IDE to build and flash SMLReader if you take care of the dependencies and the build flags configured in the `platform.io` file.
I strongly recommend using PlatformIO as it takes care of that itself.

#### Configuration

The configuration of the reading heads is done by editing `src/config.h` and adjusting  `SENSOR_CONFIGS` (see below).

```c++
static const SensorConfig SENSOR_CONFIGS[] = {
    {.pin = D2, // GPIO pin of the phototransistor
     .name = "1", // Sensor name used in MQTT topic
     .numeric_only = false, // If "true", only numeric values are being published via MQTT
     .status_led_enabled = true, // Flash status LED (3 times) when an SML start sequence has been found
     .status_led_inverted = true, // Some LEDs (like the ESP8266 builtin LED) require an inverted output signal
     .status_led_pin = LED_BUILTIN, // GPIO pin used for sensor status LED
     .interval = 0 // If greater than 0, messages are published every [interval] seconds
    },
    {.pin = D5,
     .name = "2",
     .numeric_only = false,
     .status_led_enabled = true,
     .status_led_inverted = true,
     .status_led_pin = LED_BUILTIN,
     .interval = 0
    },
    {.pin = D6,
     .name = "3",
     .numeric_only = false,
     .status_led_enabled = true,
     .status_led_inverted = true,
     .status_led_pin = LED_BUILTIN,
     .interval = 15
    }
};
```


#### Building

Building SMLReader in PlatformIO is straight forward and can be done by executing the build task matching your environment (i.e. `d1_mini`).

In case you get the following error, it is time to install a Git client (https://git-scm.com/downloads) and to make sure that the path to `git` is covered by the PATH variable and `git` is thus executable from everywhere.

```
UserSideException: Please install Git client from https://git-scm.com/downloads:
  File "/home/monty/.platformio/penv/lib/python3.8/site-packages/platformio/builder/main.py", line 168:
    env.SConscript("$BUILD_SCRIPT")
```

![Building with PlatformIO](doc/screenshots/screenshot_platformio_build.png)


#### Flashing

![Flashing with PlatformIO](doc/screenshots/screenshot_platformio_upload.png)

---


### Running

WiFi and MQTT are configured via the web interface provided by [IotWebConf](https://github.com/prampec/IotWebConf) and which can be reached after joining the WiFi network named SMLReader and heading to http://192.168.4.1.   
If the device has already been configured, the web interface can be reached via the IP address obtained from your local network's DHCP server.
To login provide the user `admin` and the configured AP password.

*Attention: You have to change the AP Password (empty by default), otherwise SMLReader won't work.*

If everything is configured properly and running with a sensor in place, SMLReader will publish the metrics and values received from the meter to the configured MQTT broker:

```
MB-Monty ➜  ~  mosquitto_sub -h 10.4.32.103 -v -t smartmeter/mains/#
smartmeter/mains/info Hello from 00C7551E, running SMLReader version 2.1.5.
smartmeter/mains/sensor/1/obis/1-0:1.8.0/255/value 3546245.9
smartmeter/mains/sensor/1/obis/1-0:2.8.0/255/value 13.2
smartmeter/mains/sensor/1/obis/1-0:1.8.1/255/value 0.0
smartmeter/mains/sensor/1/obis/1-0:2.8.1/255/value 13.2
smartmeter/mains/sensor/1/obis/1-0:1.8.2/255/value 3546245.9
smartmeter/mains/sensor/1/obis/1-0:2.8.2/255/value 0.0
smartmeter/mains/sensor/1/obis/1-0:16.7.0/255/value 451.2

smartmeter/mains/sensor/2/obis/1-0:1.8.0/255/value 3546245.9
smartmeter/mains/sensor/2/obis/1-0:2.8.0/255/value 13.2
smartmeter/mains/sensor/2/obis/1-0:1.8.1/255/value 0.0
smartmeter/mains/sensor/2/obis/1-0:2.8.1/255/value 13.2
smartmeter/mains/sensor/2/obis/1-0:1.8.2/255/value 3546245.9
smartmeter/mains/sensor/2/obis/1-0:2.8.2/255/value 0.0
smartmeter/mains/sensor/2/obis/1-0:16.7.0/255/value 451.2

smartmeter/mains/sensor/3/obis/1-0:1.8.0/255/value 3546245.9
smartmeter/mains/sensor/3/obis/1-0:2.8.0/255/value 13.2
smartmeter/mains/sensor/3/obis/1-0:1.8.1/255/value 0.0
smartmeter/mains/sensor/3/obis/1-0:2.8.1/255/value 13.2
smartmeter/mains/sensor/3/obis/1-0:1.8.2/255/value 3546245.9
smartmeter/mains/sensor/3/obis/1-0:2.8.2/255/value 0.0
smartmeter/mains/sensor/3/obis/1-0:16.7.0/255/value 451.2
```

---


### Debugging

Serial logging can be enabled by setting `SERIAL_DEBUG=true` in the `platformio.ini` file before building.
To increase the log level and to get the raw SML data, also set `SERIAL_DEBUG_VERBOSE=true`.

#### Serial port monitor

A serial port monitor can be attached using the corresponding function of your IDE or by invoking a terminal client like `miniterm` which comes shipped with `python-serial`.
The serial port settings are `115200,8,N,1`.

##### miniterm.py

```bash
miniterm /dev/ttyUSB0 115200
```

###### Example

```bash
nb-ubuntu ➜  ~/Downloads  miniterm /dev/ttyUSB0 115200
--- Miniterm on /dev/ttyUSB0  115200,8,N,1 ---
--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
Config size: 685
Wrong config version.
AP password was not set in configuration
State changing from: 0 to 1
Setting up AP: SMLReader
With default password: <hidden>
AP IP address: 192.168.4.1
State changed from: 0 to 1
...
```

Or the dockerized way:

```bash
docker run -it --device /dev/ttyUSB0 -v $(pwd):/src --rm mruettgers/esptool ash -c "miniterm.py /dev/ttyUSB0 115200"
```



##### PlatformIO

![PlatformIO Monitor](doc/screenshots/screenshot_platformio_upload_and_monitor.png)



---

## Acknowledgements

### Third party libraries
* [ESPSoftwareSerial](https://github.com/plerup/espsoftwareserial)
* [IotWebConf](https://github.com/prampec/IotWebConf)
* [MicroDebug](https://github.com/rlogiacco/MicroDebug)
* [ESPAsyncTCP](https://github.com/me-no-dev/ESPAsyncTCP)
* [Pangolin MQTT Client](https://github.com/philbowles/PangolinMQTT)
* [libSML](https://github.com/volkszaehler/libsml)
* [JLed](https://github.com/jandelgado/jled)

### Links

* https://www.msxfaq.de/sonst/bastelbude/smartmeter_d0_sml_protokoll.htm
* https://www.photovoltaikforum.com/thread/78798-sml-pr%C3%BCfsummenberechnung-ich-verzweifle-an-crc/ (Helped me to finally solve some weird CRC calculation issues before switching over to libSML)
* http://www.stefan-weigert.de/php_loader/sml.php (Sources of the 3D-printed housing of the reading head)
* https://github.com/gemu2015/Sonoff-Tasmota (Gemu's excellent Tasmota version with SML support)

## Donate

### Paypal
[![Paypal](https://www.paypalobjects.com/en_US/DK/i/btn/btn_donateCC_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=GK95YZCEGJT84)

## Roadmap

* [ ] Use LITTLEFS for config storage
* [ ] New configuration GUI based on Preact
* [ ] Configuration of sensors via web interface
* [ ] Add list of devices that are known to work
* [ ] Support for ASCII based SML messages (also known as "SML in Textform")
* [ ] Deep sleep for battery powered devices
* [ ] Grafana / InfluxDB tutorial based on docker
* [ ] KNX support for sending readings via an IP gateway to the bus

## License

Distributed under the GPL v3 license.  
See [LICENSE](LICENSE) for more information.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.3.0] - 2023-03-14
### Changed
- Upgraded IotWebConf to version 3
### Fixed
- [Issue #47](https://github.com/mruettgers/SMLReader/issues/47), Broken MQTT reconnect after WiFi disconnect

## [2.2.1] - 2022-03-17
### Changed
- [Issue #19](https://github.com/mruettgers/SMLReader/issues/19)
### Fixed
- The max length of the WiFi password has been increased to 64 chars

## [2.2.0] - 2021-05-04
### Changed
- Replaced MQTT client library with [Pangolin MQTT Client](https://github.com/philbowles/PangolinMQTT)
- Refactored throttling logic
### Fixed
- [Issue #15](https://github.com/mruettgers/SMLReader/issues/15)

## [2.1.6] - 2021-01-03
### Added
- Configuration option to allow setting an inverval for throttling the MQTT messages
### Fixed
- Locked third party dependency IotWebConf to version 2 due to incompatible changes in version 3

## [2.1.5] - 2020-07-25
### Changed
- Modified MQTT topic to use / instead of * due to invalid character issues when used with ioBroker

## [2.1.4] - 2020-07-24
### Changed
- Worked on the docs
- Added LED based feedback for SML data recognition

## [2.1.3] - 2020-04-17
### Changed
- Added hint to change AP password to the docs
- Moved WiFi default settings to `config.h`
### Fixed
- String initialization issues resulting in random MQTT username/passqord in UI

## [2.1.2] - 2020-04-16
### Changed
- Updated the docs
- Read (and in debug mode print out) sensor values even without an active wifi connection

## [2.1.1] - 2020-04-15
### Changed
- Fixed MQTT topic to include the whole OBIS identifier
- Look for a reading head attached to GPIO pin D2 by default
- Switched to official libSML repository


src/unit.h
/**
 * DLMS Units as specified in ISO EN 62056-62 and used by SML
 *
 * @package libsml
 * @copyright Copyright (c) 2011, The volkszaehler.org project
 * @license http://www.gnu.org/licenses/gpl.txt GNU Public License
 * @author Steffen Vogel <info@steffenvogel.de>
 */
/*
 * This file is part of volkzaehler.org
 *
 * volkzaehler.org is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * volkzaehler.org is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with volkszaehler.org. If not, see <http://www.gnu.org/licenses/>.
 */

typedef struct {
	unsigned char code;
	const char *unit;
} dlms_unit_t;

/**
 * Static lookup table
 */
dlms_unit_t dlms_units[] = {
// code, unit		// Quantity			Unit name		SI definition (comment)
//=====================================================================================================
{1, "a"},		// time				year			52*7*24*60*60 s
{2, "mo"},		// time				month			31*24*60*60 s
{3, "wk"},		// time				week			7*24*60*60 s
{4, "d"},		// time				day			24*60*60 s
{5, "h"},		// time				hour			60*60 s
{6, "min."},		// time				min			60 s
{7, "s"},		// time (t)			second			s
{8, "°"},		// (phase) angle		degree			rad*180/π
{9, "°C"},		// temperature (T)		degree celsius		K-273.15
{10, "currency"},	// (local) currency
{11, "m"},		// length (l)			metre			m
{12, "m/s"},		// speed (v)			metre per second	m/s
{13, "m³"},		// volume (V)			cubic metre		m³
{14, "m³"},		// corrected volume		cubic metre		m³
{15, "m³/h"},		// volume flux			cubic metre per hour 	m³/(60*60s)
{16, "m³/h"},		// corrected volume flux	cubic metre per hour 	m³/(60*60s)
{17, "m³/d"},		// volume flux						m³/(24*60*60s)
{18, "m³/d"},		// corrected volume flux				m³/(24*60*60s)
{19, "l"},		// volume			litre			10-3 m³
{20, "kg"},		// mass (m)			kilogram
{21, "N"},		// force (F)			newton
{22, "Nm"},		// energy			newtonmeter		J = Nm = Ws
{23, "Pa"},		// pressure (p)			pascal			N/m²
{24, "bar"},		// pressure (p)			bar			10⁵ N/m²
{25, "J"},		// energy			joule			J = Nm = Ws
{26, "J/h"},		// thermal power		joule per hour		J/(60*60s)
{27, "W"},		// active power (P)		watt			W = J/s
{28, "VA"},		// apparent power (S)		volt-ampere
{29, "var"},		// reactive power (Q)		var
{30, "Wh"},		// active energy		watt-hour		W*(60*60s)
{31, "VAh"},		// apparent energy		volt-ampere-hour	VA*(60*60s)
{32, "varh"},		// reactive energy		var-hour		var*(60*60s)
{33, "A"},		// current (I)			ampere			A
{34, "C"},		// electrical charge (Q)	coulomb			C = As
{35, "V"},		// voltage (U)			volt			V
{36, "V/m"},		// electr. field strength (E)	volt per metre
{37, "F"},		// capacitance (C)		farad			C/V = As/V
{38, "Ω"},		// resistance (R)		ohm			Ω = V/A
{39, "Ωm²/m"},		// resistivity (ρ)		Ωm
{40, "Wb"},		// magnetic flux (Φ)		weber			Wb = Vs
{41, "T"},		// magnetic flux density (B)	tesla			Wb/m2
{42, "A/m"},		// magnetic field strength (H)	ampere per metre	A/m
{43, "H"},		// inductance (L)		henry			H = Wb/A
{44, "Hz"},		// frequency (f, ω)		hertz			1/s
{45, "1/(Wh)"},		// R_W							(Active energy meter constant or pulse value)
{46, "1/(varh)"},	// R_B							(reactive energy meter constant or pulse value)
{47, "1/(VAh)"},	// R_S							(apparent energy meter constant or pulse value)
{48, "V²h"},		// volt-squared hour		volt-squaredhours	V²(60*60s)
{49, "A²h"},		// ampere-squared hour		ampere-squaredhours	A²(60*60s)
{50, "kg/s"},		// mass flux			kilogram per second	kg/s
{51, "S, mho"},		// conductance siemens					1/Ω
{52, "K"},		// temperature (T)		kelvin
{53, "1/(V²h)"},	// R_U²h						(Volt-squared hour meter constant or pulse value)
{54, "1/(A²h)"},	// R_I²h						(Ampere-squared hour meter constant or pulse value)
{55, "1/m³"},		// R_V, meter constant or pulse value (volume)
{56, "%"},		// percentage			%
{57, "Ah"},		// ampere-hours			ampere-hour
{60, "Wh/m³"},		// energy per volume					3,6*103 J/m³
{61, "J/m³"},		// calorific value, wobbe
{62, "Mol %"},		// molar fraction of		mole percent		(Basic gas composition unit)
			// gas composition  
{63, "g/m³"},		// mass density, quantity of material			(Gas analysis, accompanying elements)
{64, "Pa s"},		// dynamic viscosity pascal second			(Characteristic of gas stream)
{253, "(reserved)"},	// reserved
{254, "(other)"},	// other unit
{255, "(unitless)"},	// no unit, unitless, count
{0, ""}		// stop condition for iterator
};
	
const char * dlms_get_unit(unsigned char code) {
	dlms_unit_t *it = dlms_units;
	do { // linear search
		if (it->code == code) {
			return it->unit;
		}
	} while ((++it)->code);
	
	return NULL; // not found
}